cmd/iptables-wrapper.go

```diff
@@ -0,0 +1,174 @@
+	iptablesNFTMode    string = "nft"
+	iptablesLegacyMode string = "legacy"
+)
+
+// TODO: Move this comment to a better place
+/*
+Assuming that when assembling the container the following binaries will be linked to
+this wrapper:
+* iptables
+* iptables-wrapper
+* iptables-save
+* iptables-restore
+* ip6tables
+* ip6tables-save
+* ip6tables-restore
+
+1) This binary will try to detect which iptables mode was being used by kubelet,
+calling `xtables-<mode>-multi` and checking if the kubelet rules exists in this path
+2) Determined the right mode/binary, it will call `xtables-<mode>-multi` passing
+the command and the arguments to it.
+*/
+func main() {
+	if err := checkIptablesNFTVersion(); err != nil {
+		fmt.Fprintf(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	iptablesMode, err := getIptablesMode()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	var iptablesBinary string
+
+	switch iptablesMode {
+	case iptablesNFTMode:
+		iptablesBinary = iptablesNFTPath
+	case iptablesLegacyMode:
+		iptablesBinary = iptablesLegacyPath
+	default:
+		fmt.Fprintf(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	args := make([]string, 0)
+	args = append(args, filepath.Base(os.Args[0]))
+	if len(os.Args) > 0 {
+		args = append(args, os.Args...)
+	}
+
+	// Buffers are faster than standard string reading
+	var output bytes.Buffer
+
+	cmdIptables := &exec.Cmd{
+		Path:   iptablesBinary,
+		Args:   args,
+		Stdout: &output,
+	}
+
+	if err := cmdIptables.Run(); err != nil {
+		fmt.Fprintf(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	fmt.Print(output.String())
+
+}
+
+func checkIptablesNFTVersion() error {
+	outNft, err := exec.Command(iptablesNFTPath, "iptables", "--version").Output()
+	if err != nil {
+		return err
+	}
+	if regexIPTVers.Match(outNft) {
+		return errors.New("ERROR: iptables 1.8.0 - 1.8.3 have compatibility bugs. Upgrade to v1.8.4 or newer.")
+	}
+	return nil
+}
+
+// getIptablesMode uses the tries to find the pre-existing "KUBE-IPTABLES-HINT"
+// or "KUBE-KUBELET-CANARY" to determine which iptables mode should be used
+func getIptablesMode() (string, error) {
+	/*
+		In kubernetes 1.17 and later, kubelet will have created at least
+		one chain in the "mangle" table (either "KUBE-IPTABLES-HINT" or
+		"KUBE-KUBELET-CANARY"), so check that first, against
+		iptables-nft, because we can check that more efficiently and
+		it's more common these days.
+	*/
+
+	outNft, err := exec.Command(iptablesNFTPath, "iptables-save",
+		"-t", "mangle").Output()
+
+	if err != nil {
+		return "", err
+	}
+
+	outNftv6, err := exec.Command(iptablesNFTPath, "ip6tables-nft-save",
+		"-t", "mangle").Output()
+
+	if err != nil {
+		return "", err
+	}
+
+	if hasKubernetesChains(outNft) || hasKubernetesChains(outNftv6) {
+		return iptablesNFTMode, nil
+	}
+
+	// Per @danwinship comment, we can deprecate the logic of getting NFT vs Legacy
+	// rules and assume if it's not NFT, then it's legacy and the rules should have
+	// been created in Legacy
```

Comment:
This comment only makes sense if you know that the shell script version does something different.

You also dropped the comment from the shell script version that explains why we don't use "-t" here:

```
    # Check for kubernetes 1.17-or-later with iptables-legacy. We
    # can't pass "-t mangle" to iptables-legacy-save because it would
    # cause the kernel to create that table if it didn't already
    # exist, which we don't want. So we have to grab all the rules
```

I'd put that here, and then at the end of the function you can have a comment about "the shell script version has further checks here for older versions of kubernetes, but the binary version doesn't support that" or something